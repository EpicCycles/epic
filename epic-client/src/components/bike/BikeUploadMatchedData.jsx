import React from "react";
import {Icon} from "semantic-ui-react";

export const BikeUploadMatchedData = props => (
    <div
        key={`match${props.matched.rowIndex}${props.matchIndex}`}
        className={(props.matchIndex > 1) ? "rounded-auto red" : "rounded-auto "}
    >
        {props.matched.rowField}
        <Icon
            key={`matchDelete$${props.matched.rowIndex}${props.matchIndex}`}
            name="delete"
            onClick={() => props.undoMapping(props.matched.rowIndex)}
        />
    </div>
);