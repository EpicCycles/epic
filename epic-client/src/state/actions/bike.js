export const CLEAR_FRAME = "bike/CLEAR_FRAME";
export const BIKE_REVIEW_REQUESTED = "bike/BIKE_REVIEW_REQUESTED";
export const BIKE_REVIEW_BIKE = "bike/BIKE_REVIEW_BIKE";
export const BIKE_DELETE_REQUESTED = "bike/BIKE_DELETE_REQUESTED";
export const BIKE_DELETE_OK = "bike/BIKE_DELETE_OK";
export const BIKE_DELETE_ERROR = "bike/BIKE_DELETE_ERROR";
export const BIKE_ADD_PART_REQUESTED = "bike/BIKE_ADD_PART_REQUESTED";
export const BIKE_ADD_PART_OK = "bike/BIKE_ADD_PART_OK";
export const BIKE_ADD_PART_ERROR = "bike/BIKE_ADD_PART_ERROR";
export const BIKE_SAVE_REQUESTED = "bike/BIKE_SAVE_REQUESTED";
export const BIKE_SAVE_OK = "bike/BIKE_SAVE_OK";
export const BIKE_SAVE_ERROR = "bike/BIKE_SAVE_ERROR";
export const BIKE_PART_DELETE_REQUESTED = "bike/BIKE_PART_DELETE_REQUESTED";
export const BIKE_PART_DELETE_OK = "bike/BIKE_PART_DELETE_OK";
export const BIKE_PART_DELETE_ERROR = "bike/BIKE_PART_DELETE_ERROR";
export const BIKE_PART_SAVE_REQUESTED = "bike/BIKE_PART_SAVE_REQUESTED";
export const BIKE_PART_SAVE_OK = "bike/BIKE_PART_SAVE_OK";
export const BIKE_PART_SAVE_ERROR = "bike/BIKE_PART_SAVE_ERROR";
export const FRAME_ARCHIVE_REQUESTED = "bike/FRAME_ARCHIVE_REQUESTED";
export const FRAME_ARCHIVE_OK = "bike/FRAME_ARCHIVE_OK";
export const FRAME_ARCHIVE_ERROR = "bike/FRAME_ARCHIVE_ERROR";
export const FRAME_DELETE_REQUESTED = "bike/FRAME_DELETE_REQUESTED";
export const FRAME_DELETE_OK = "bike/FRAME_DELETE_OK";
export const FRAME_DELETE_ERROR = "bike/FRAME_DELETE_ERROR";
export const FRAME_SAVE_REQUESTED = "bike/FRAME_SAVE_REQUESTED";
export const FRAME_SAVE_OK = "bike/FRAME_SAVE_OK";
export const FRAME_SAVE_ERROR = "bike/FRAME_SAVE_ERROR";
export const FRAME_UPLOAD_REQUESTED = "bike/FRAME_UPLOAD_REQUESTED";
export const FRAME_UPLOAD_OK = "bike/FRAME_UPLOAD_OK";
export const FRAME_UPLOAD_ERROR = "bike/FRAME_UPLOAD_ERROR";
export const FRAME_LIST_REQUESTED = "bike/FRAME_LIST_REQUESTED";
export const FRAME_LIST_OK = "bike/FRAME_LIST_OK";
export const FRAME_LIST_ERROR = "bike/FRAME_LIST_ERROR";


export const clearFrame = () => ({
    type: CLEAR_FRAME
});
export const saveBike = (bike) => ({
    type: BIKE_SAVE_REQUESTED,
    payload: { bike }
});
export const saveBikeOK = (bike) => ({
    type: BIKE_SAVE_OK,
    payload: { bike }
});
export const saveBikeError = (error) => ({
    type: BIKE_SAVE_ERROR,
    payload: error
});
export const saveBikePart = (bikeId, part) => ({
    type: BIKE_PART_SAVE_REQUESTED,
    payload: { bikeId, part }
});
export const saveBikePartOK = (parts) => ({
    type: BIKE_PART_SAVE_OK,
    payload: { parts }
});
export const saveBikePartError = (error) => ({
    type: BIKE_PART_SAVE_ERROR,
    payload: error
});
export const addBikePart = (bikeId, part) => ({
    type: BIKE_ADD_PART_REQUESTED,
    payload: { bikeId, part }
});
export const addBikePartOK = (parts) => ({
    type: BIKE_ADD_PART_OK,
    payload: { parts }
});
export const addBikePartError = (error) => ({
    type: BIKE_ADD_PART_ERROR,
    payload: error
});
export const deleteBikePart = (bikeId, partId) => ({
    type: BIKE_PART_DELETE_REQUESTED,
    payload: { bikeId, partId }
});
export const deleteBikePartOK = (parts) => ({
    type: BIKE_PART_DELETE_OK,
    payload: { parts }
});
export const deleteBikePartError = (error) => ({
    type: BIKE_PART_DELETE_ERROR,
    payload: error
});
export const reviewBikes = (bikeReviewList) => ({
    type:BIKE_REVIEW_REQUESTED,
    payload: {bikeReviewList}
});
export const reviewBike = (bikeId) => ({
    type:BIKE_REVIEW_BIKE,
    payload: { bikeId }
});

export const deleteBikes = (bikeDeleteList, searchCriteria) => ({
    type: BIKE_DELETE_REQUESTED,
    payload: { bikeDeleteList, searchCriteria }
});
export const deleteBikesError = (error) => ({
    type: BIKE_DELETE_ERROR,
    payload: error
});
export const deleteBikesSuccess = () => ({
    type: BIKE_DELETE_OK
});
export const archiveFrames = (frameArchiveList, searchCriteria) => ({
    type: FRAME_ARCHIVE_REQUESTED,
    payload: { frameArchiveList, searchCriteria }
});
export const archiveFramesError = (error) => ({
    type: FRAME_ARCHIVE_ERROR,
    payload: error
});
export const archiveFramesSuccess = () => ({
    type: FRAME_ARCHIVE_OK
});
export const deleteFrames = (frameDeleteList, searchCriteria) => ({
    type: FRAME_DELETE_REQUESTED,
    payload: { frameDeleteList, searchCriteria }
});
export const deleteFramesError = (error) => ({
    type: FRAME_DELETE_ERROR,
    payload: error
});
export const deleteFramesSuccess = () => ({
    type: FRAME_DELETE_OK
});
export const saveFrame = (frame, searchCriteria) => ({
    type: FRAME_SAVE_REQUESTED,
    payload: { frame, searchCriteria }
});
export const saveFrameError = (error) => ({
    type: FRAME_SAVE_ERROR,
    payload: error
});
export const saveFrameSuccess = (frame) => ({
    type: FRAME_SAVE_OK,
    payload: frame
});
export const uploadFrame = (frame) => ({
    type: FRAME_UPLOAD_REQUESTED,
    payload: { frame }
});
export const uploadFrameError = (error) => ({
    type: FRAME_UPLOAD_ERROR,
    payload: error
});
export const uploadFrameSuccess = () => ({
    type: FRAME_UPLOAD_OK,
});
export const getFrameList = (searchCriteria) => ({
    type: FRAME_LIST_REQUESTED,
    payload: searchCriteria
});
export const getFrameListOK = (apiData) => ({
    type: FRAME_LIST_OK,
    payload: apiData
});
export const getFrameListError = (error) => ({
    type: FRAME_LIST_ERROR,
    payload: error
});
