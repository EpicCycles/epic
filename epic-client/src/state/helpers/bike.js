export const replaceBikeParts = (bikeId, newBikeParts, existingBikeParts) => {
    let bikePartsForState = existingBikeParts.filter(oldBikePart => oldBikePart.bike !== bikeId);
    return bikePartsForState.concat(newBikeParts);
};