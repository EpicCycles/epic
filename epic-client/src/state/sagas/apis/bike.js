import api from '../api';

const saveFrame = async (payload) => {
    api.instance.defaults.headers.common['Authorization'] = `Token ${payload.token}`;
    return await api.instance.patch(`api/frames/${payload.frame.id}`, payload.frame);
};

const uploadFrame = async (payload) => {
    api.instance.defaults.headers.common['Authorization'] = `Token ${payload.token}`;
    return await api.instance.post("api/frame/upload/", payload.frame);
};

const getFrames = async (payload) => {
    api.instance.defaults.headers.common['Authorization'] = `Token ${payload.token}`;
    const frameApi = `/api/frames?brand=${payload.brand}&frameName=${payload.frameName}&archived=${payload.archived}`;
    return await api.instance.get(frameApi);
};

export default {
    uploadFrame,
    getFrames,
    saveFrame,
}
