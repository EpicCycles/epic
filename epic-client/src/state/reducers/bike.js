import {CLEAR_ALL_STATE} from "../actions/application";
import {
    BIKE_ADD_PART_ERROR,
    BIKE_ADD_PART_OK,
    BIKE_ADD_PART_REQUESTED,
    BIKE_DELETE_ERROR,
    BIKE_PART_DELETE_ERROR,
    BIKE_PART_DELETE_OK,
    BIKE_PART_DELETE_REQUESTED,
    BIKE_PART_SAVE_ERROR,
    BIKE_PART_SAVE_OK,
    BIKE_PART_SAVE_REQUESTED,
    BIKE_REVIEW_BIKE,
    BIKE_REVIEW_ERROR,
    BIKE_REVIEW_OK,
    BIKE_REVIEW_REQUESTED,
    BIKE_SAVE_ERROR,
    BIKE_SAVE_OK,
    BIKE_SAVE_REQUESTED,
    CLEAR_FRAME,
    FRAME_ARCHIVE_ERROR,
    FRAME_DELETE_ERROR,
    FRAME_LIST_ERROR,
    FRAME_LIST_OK,
    FRAME_LIST_REQUESTED,
    FRAME_SAVE_ERROR,
    FRAME_SAVE_OK,
    FRAME_SAVE_REQUESTED,
    FRAME_UPLOAD_ERROR,
    FRAME_UPLOAD_OK,
    FRAME_UPLOAD_REQUESTED
} from "../actions/bike";

const initialState = {
    isLoading: false
};

const bike = (state = initialState, action) => {
    switch (action.type) {
        case CLEAR_ALL_STATE:
        case CLEAR_FRAME:
            return initialState;
        case BIKE_SAVE_REQUESTED:
        case BIKE_PART_DELETE_REQUESTED:
        case BIKE_PART_SAVE_REQUESTED:
        case BIKE_ADD_PART_REQUESTED:
            return {
                ...state,
                isLoading: true
            };
        case BIKE_SAVE_OK:
            return {
                ...state,
                bike: action.payload.bike,
                isLoading: false
            };
        case BIKE_PART_SAVE_OK:
        case BIKE_PART_DELETE_OK:
        case BIKE_ADD_PART_OK:
            return {
                ...state,
                parts: action.payload.parts,
                isLoading: false
            };
        case BIKE_REVIEW_REQUESTED:
            // state should just be the list of bikes
            return {
                bikeReviewList: action.payload.bikeReviewList,
                isLoading: true
            };
        case BIKE_REVIEW_BIKE:
            return {
                ...state,
                bikeId: action.payload.bikeId,
                bike: {},
                parts: [],
                isLoading: true
            };
        case BIKE_REVIEW_OK:
            return {
                ...state,
                bike: action.payload.bike,
                parts: action.payload.parts,
                isLoading: false
            };
        case FRAME_SAVE_REQUESTED:
        case FRAME_UPLOAD_REQUESTED:
            return {
                ...state,
                frame: action.payload.frame,
                isLoading: true,
            };
        case FRAME_SAVE_OK:
            return {
                ...state,
                frame: action.payload,
                isLoading: false,
            };
        case FRAME_UPLOAD_OK:
            return {
                ...state,
                isLoading: false,
            };
        case FRAME_LIST_REQUESTED:
            return {
                ...state,
                isLoading: true,
            };
        case FRAME_LIST_OK:
            return {
                ...state,
                frames: action.payload,
                isLoading: false,
            };
        case BIKE_REVIEW_ERROR:
        case BIKE_DELETE_ERROR:
        case BIKE_SAVE_ERROR:
        case BIKE_ADD_PART_ERROR:
        case BIKE_PART_SAVE_ERROR:
        case BIKE_PART_DELETE_ERROR:
        case FRAME_SAVE_ERROR:
        case FRAME_UPLOAD_ERROR:
        case FRAME_LIST_ERROR:
        case FRAME_ARCHIVE_ERROR:
        case FRAME_DELETE_ERROR:
            return {
                ...state,
                isLoading: false,
            };
        default:
            return state;
    }
};

export default bike;