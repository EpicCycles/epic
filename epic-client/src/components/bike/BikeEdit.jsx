import React, {Fragment} from "react";
import * as PropTypes from "prop-types";

import {getUpdatedObject, isItAnObject, updateObject} from "../../helpers/utils";
import {addFieldToState} from "../app/model/helpers/model";
import {bikeFields} from "../app/model/helpers/fields";
import {validateData} from "../app/model/helpers/validators";
import {Icon} from "semantic-ui-react";
import EditModelPage from "../app/model/EditModelPage";
import {bikeFullName} from "./helpers/bike_helpers";

class BikeEdit extends React.Component {
    state = {};

    componentWillMount() {
        this.setState(this.deriveStateFromProps());
    };


    deriveStateFromProps = () => {
        let newState = updateObject(this.props.bike)
        return newState;
    };

    handleInputChange = (fieldName, input) => {
        let newState = addFieldToState(this.state, bikeFields, fieldName, input);

        if (this.checkForChanges(newState)) {
            newState.errors = validateData(bikeFields, newState);
        }

        this.checkBikeDataList(newState);
        this.setState(newState);
    };

    onClickReset = () => {
        const resetState = this.deriveStateFromProps();
        this.setState(resetState);
        this.checkBikeDataList(resetState);
    };

    saveOrCreateBike = () => {
        const updatedBike = getUpdatedObject(bikeFields, this.props.bike, this.state);
        this.props.saveBike(updatedBike);
        if (this.props.closeModal) {
            this.props.closeModal();
        }
    };
    deleteOrRemoveBike = () => {
        if (this.state.id) {
            this.props.deleteBike(this.state.id);
        }
        this.setState({});
        if (this.props.closeModal) {
            this.props.closeModal();
        }
    };

    render() {
        const { bike, brands, frames } = this.props;
        const {changed} = this.state;

        return <Fragment>
            <h3>{bikeFullName(bike, frames, brands)}</h3>
            <EditModelPage
                model={this.state}
                modelFields={bikeFields}
                onChange={this.handleInputChange}
                persistedModel={bike}
            />
            <div style={{ width: "100%", textAlign: "right" }}>
                {changed &&
                <Icon id={`reset-bike`} name="undo"
                      onClick={this.onClickReset}
                      title="Reset Bike details"
                />
                }
                {(isChanged && !isItAnObject(errors)) &&
                <Icon id={`accept-bike`} name="check"
                      onClick={this.saveOrCreateBike}
                      title="Confirm Bike Change"
                />
                }
                {deleteBike &&
                <Icon id={`delete-bike`} name="trash"
                      onClick={this.deleteOrRemoveBike}
                      title="Delete Bike"
                />
                }
            </div>
        </Fragment>;
    }
}
BikeEdit.propTypes = {
    bike: PropTypes.object.isRequired,
    brands: PropTypes.array.isRequired,
    frames: PropTypes.array.isRequired,
    saveBike: PropTypes.func.isRequired,
    deleteBike: PropTypes.func.isRequired,
};
export default BikeEdit;