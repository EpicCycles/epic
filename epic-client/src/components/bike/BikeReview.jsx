import React, {Fragment} from 'react'
import {Dimmer, Loader} from "semantic-ui-react";
import Pagination from "../../common/pagination";
import {updateObject} from "../../helpers/utils";
import {Redirect} from "react-router";
import {applyFieldValueToModel} from "../app/model/helpers/model";
import BikeEdit from "./BikeEdit";
import * as PropTypes from "prop-types";
import {findPartsForBike} from "./helpers/bike";
import PartDisplayGrid from "../part/PartDisplayGrid";
import PartFinder from "../part/PartFinder";

const initialState = {
    showPartFinder: false,
    partEditPart:{},
}
// Review page for a single bike - needs: bike field entry section, table of parts with edit ability for each part

// TODO revise with new api structure
// 1.  state - check prev props and bikeId - get details for that bike id onload, frame name from frames.
// 2.  remove modal for parts - instead add panel to add a new part
// 3.  when a part is saved the api call gets bike parts and parts - needs revising
//     - api change - return bikeparts, parts
//     - reducer change to remove all bike parts for bike and add values from api - and trigger parts being added
// 4.  when a bike is deleted - api OK (I thnk) - check that bike and bike parts are removed and that bike id for review is updated, and id removed from review array
// 5.  next review - check that id is updated
// 6. display - drive from sections in body rather than in initialisation
// 7.  allow edit part details (price trade in price etc) as well as use different part - different part will be search for parts or enter new part block
class BikeReview extends React.Component {
    constructor() {
        super();
    }

    componentDidMount() {
        this.checkPropsData();
    };

    componentDidUpdate(prevProps) {
        this.checkPropsData();
        if (prevProps.bikeId !== this.props.bikeId) this.setState(initialState);
    };

    checkPropsData = () => {
        if (!this.props.isLoading) {
            this.getData();
        }
    };
    getData = () => {
        let brandsRequired = true;
        let frameworkRequired = true;
        if (this.props.brands && this.props.brands.length > 0) {
            brandsRequired = false;
        }
        if (this.props.sections && this.props.sections.length > 0) {
            frameworkRequired = false;
        }
        if (brandsRequired) {
            this.props.getBrandsAndSuppliers();
        } else if (frameworkRequired) {
            this.props.getFramework();
        }
    };
    reviewSelectedBike = (bikePage) => {
        const bikeIndex = bikePage - 1;
        this.props.reviewBike(this.props.bikeReviewList[bikeIndex]);
    };
    deletePart = (partId) => {
        this.props.deleteBikePart(this.props.bike.id, partId);
    };
    saveOrAddPart = (part) => {
        const bikeId = this.props.bike.id;
        if (part.id) {
            this.props.saveBikePart(bikeId, part);
        } else {
            this.props.addBikePart(bikeId, part);
        }
    };
    showPartFinder = (part) => {
        this.setState({partEditPart: part, partFinder: true});
    };
    closePartFinder = (part) => {
        this.setState(initialState);
    };
    render() {
        const {bikes, bikeParts, parts, bikeReviewList, isLoading, brands, frames, sections, saveBike, deleteBike, bikeId, listParts} = this.props;
        const {partEditPart, partFinder} = this.state;
        const selectedBikeIndex = bikeReviewList.indexOf(bikeId);
        const bike = selectedBikeIndex ? bikes[selectedBikeIndex] : undefined;
        const partsForBike = findPartsForBike(bike, bikeParts, parts);
        return <Fragment key={`bikeReview`}>
            {!(bike) && <Redirect to="/bike-review-list" push/>}
            <section className="row">
                {partFinder && <PartFinder
                    sections={sections}
                    brands={brands}
                    savePart={this.saveOrAddPart}
                    deletePart={this.deletePart}
                    findParts={listParts}
                    part={partEditPart}
                    closeAction={this.closePartFinder}
                    partActionPrimary={this.saveOrAddPart}
                    partActionPrimaryIcon={'add'}
                    partActionPrimaryTitle={'Update bike with part'}
                />}
            </section>
            <BikeEdit
                bike={bike}
                brands={brands}
                frames={frames}
                saveBike={saveBike}
                deleteBike={deleteBike}
            />
            <PartDisplayGrid
                parts={partsForBike}
                sections={sections}
                brands={brands}
                editPart={this.showPartFinder}
            />
            <Pagination
                type="Bike"
                getPage={this.reviewSelectedBike}
                lastPage={bikeReviewList.length}
                count={bikeReviewList.length}
                page={(selectedBikeIndex + 1)}
            />
            {isLoading &&
            <Dimmer active inverted>
                <Loader content='Loading'/>
            </Dimmer>
            }
        </Fragment>
    }
}

BikeReview.propTypes = {
    bikeId: PropTypes.object.isRequired,
    bikeReviewList: PropTypes.array.isRequired,
    bikeParts: PropTypes.array.isRequired,
    brands: PropTypes.array.isRequired,
    sections: PropTypes.array.isRequired,
    parts: PropTypes.array.isRequired,
    frames: PropTypes.array.isRequired,
    saveBike: PropTypes.func.isRequired,
    deleteBike: PropTypes.func.isRequired,
    reviewBike: PropTypes.func.isRequired,
    addBikePart: PropTypes.func.isRequired,
    saveBikePart: PropTypes.func.isRequired,
    deleteBikePart: PropTypes.func.isRequired,
    listParts: PropTypes.func.isRequired,
    isLoading: PropTypes.bool,
};
export default BikeReview;

