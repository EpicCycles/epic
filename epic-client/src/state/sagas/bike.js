import {call, put, select, takeLatest} from "redux-saga/effects";
import * as selectors from "../selectors/user";
import bike from "./apis/bike";
import history from "../../history";
import {
    archiveFramesError,
    archiveFramesSuccess,
    BIKE_DELETE_REQUESTED, BIKE_REVIEW_BIKE, BIKE_REVIEW_REQUESTED,
    deleteBikesError,
    deleteBikesSuccess,
    deleteFramesError,
    deleteFramesSuccess,
    FRAME_ARCHIVE_REQUESTED,
    FRAME_DELETE_REQUESTED,
    FRAME_LIST_REQUESTED,
    FRAME_SAVE_REQUESTED,
    FRAME_UPLOAD_REQUESTED,
    getFrameListError,
    getFrameListOK, reviewBikeError, reviewBikeOK,
    saveFrameError,
    saveFrameSuccess,
    uploadFrameError,
    uploadFrameSuccess,
    reviewBike,
} from "../actions/bike";

export function* reviewBikeStart(action) {
    const bikeIdsToReview = action.payload.bikeReviewList;
    if (bikeIdsToReview && bikeIdsToReview.length > 0) {
        yield put(reviewBike(bikeIdsToReview[0]));
        yield call(history.push, "/bike-browse");
    } else{
        yield put(reviewBikeError("No Bikes to Review"));
    }

}

export function* watchForReviewBikeStart() {
    yield takeLatest(BIKE_REVIEW_REQUESTED, reviewBikeStart);
}

export function* reviewBikeParts(action) {
    try {
        const token = yield select(selectors.token);
        if (token) {
            const completePayload = Object.assign(action.payload, { token });
            const response = yield call(bike.getBikeParts, completePayload);
            yield put(reviewBikeOK(response.data));
        } else {
            yield call(history.push, "/login");
        }
    } catch (error) {
        yield put(reviewBikeError("Get Bike for Review failed"));
    }
}

export function* watchForReviewBike() {
    yield takeLatest(BIKE_REVIEW_BIKE, reviewBikeParts);
}

export function* deleteBikes(bikeIdsToDelete, token) {
    try {
        for (let i = 0; i < bikeIdsToDelete.length; i++) {
            yield call(bike.deleteBike, { bikeId: bikeIdsToDelete[i], token });
        }

    } catch (error) {
        yield put(deleteBikesError("Delete Frames failed"));
    }
}

export function* deleteBikesAndGetList(action) {
    try {
        const token = yield select(selectors.token);
        if (token) {
            const bikeIdsToDelete = action.payload.bikeDeleteList;
            const searchCriteria = action.payload.searchCriteria;

            if (bikeIdsToDelete && bikeIdsToDelete.length > 0) {
                yield* deleteBikes(bikeIdsToDelete, token);
            }

            if (searchCriteria && searchCriteria.brand) {
                const searchPayload = Object.assign(searchCriteria, { token });
                const searchResponse = yield call(bike.getFrames, searchPayload);
                yield put(getFrameListOK(searchResponse.data));
            } else {
                yield put(deleteBikesSuccess());
            }
        } else {
            yield call(history.push, "/login");
        }
    } catch (error) {
        yield put(deleteBikesError("Delete Bikes failed"));
    }
}

export function* watchForDeleteBikes() {
    yield takeLatest(BIKE_DELETE_REQUESTED, deleteBikesAndGetList);
}

export function* deleteFrames(frameIdsToDelete, token) {
    try {
        for (let i = 0; i < frameIdsToDelete.length; i++) {
            yield call(bike.deleteFrame, { frameId: frameIdsToDelete[i], token });
        }

    } catch (error) {
        yield put(deleteFramesError("Delete Frames failed"));
    }
}

export function* deleteFramesAndGetList(action) {
    try {
        const token = yield select(selectors.token);
        if (token) {
            const frameIdsToDelete = action.payload.frameDeleteList;
            const searchCriteria = action.payload.searchCriteria;

            if (frameIdsToDelete && frameIdsToDelete.length > 0) {
                yield* deleteFrames(frameIdsToDelete, token);
            }

            if (searchCriteria && searchCriteria.brand) {
                const searchPayload = Object.assign(searchCriteria, { token });
                const searchResponse = yield call(bike.getFrames, searchPayload);
                yield put(getFrameListOK(searchResponse.data));
            } else {
                yield put(deleteFramesSuccess());
            }
        } else {
            yield call(history.push, "/login");
        }
    } catch (error) {
        yield put(deleteFramesError("Delete Frames failed"));
    }
}

export function* watchForDeleteFrames() {
    yield takeLatest(FRAME_DELETE_REQUESTED, deleteFramesAndGetList);
}
export function* archiveFrames(frameIdsToArchive, token) {
    try {
        for (let i = 0; i < frameIdsToArchive.length; i++) {
            const frame = { id: frameIdsToArchive[i], archived: true };
            yield call(bike.saveFrame, { frame, token });
        }

    } catch (error) {
        yield put(archiveFramesError("Archive Frames failed"));
    }
}

export function* archiveFramesAndGetList(action) {
    try {
        const token = yield select(selectors.token);
        if (token) {
            const frameIdsToArchive = action.payload.frameArchiveList;
            const searchCriteria = action.payload.searchCriteria;

            if (frameIdsToArchive && frameIdsToArchive.length > 0) {
                yield* archiveFrames(frameIdsToArchive, token);
            }

            if (searchCriteria && searchCriteria.brand) {
                const searchPayload = Object.assign(searchCriteria, { token });
                const searchResponse = yield call(bike.getFrames, searchPayload);
                yield put(getFrameListOK(searchResponse.data));
            } else {
                yield put(archiveFramesSuccess());
            }
        } else {
            yield call(history.push, "/login");
        }
    } catch (error) {
        yield put(archiveFramesError("Archive Frames failed"));
    }
}

export function* watchForArchiveFrames() {
    yield takeLatest(FRAME_ARCHIVE_REQUESTED, archiveFramesAndGetList);
}

export function* saveFrame(action) {
    try {
        const token = yield select(selectors.token);
        if (token) {
            const completePayload = Object.assign(action.payload, { token });
            const response = yield call(bike.saveFrame, completePayload);
            if (completePayload.searchCriteria) {
                const searchPayload = Object.assign(completePayload.searchCriteria, { token });
                const searchResponse = yield call(bike.getFrames, searchPayload);
                yield put(getFrameListOK(searchResponse.data));
            } else {
                yield put(saveFrameSuccess(response.data));
            }
        } else {
            yield call(history.push, "/login");
        }
    } catch (error) {
        yield put(saveFrameError("Save Frame failed"));
    }
}

export function* watchForSaveFrame() {
    yield takeLatest(FRAME_SAVE_REQUESTED, saveFrame);
}

export function* uploadFrame(action) {
    try {
        const token = yield select(selectors.token);
        if (token) {
            const completePayload = Object.assign(action.payload, { token });
            const response = yield call(bike.uploadFrame, completePayload);
            yield put(uploadFrameSuccess(response.data));
        } else {
            yield call(history.push, "/login");
        }
    } catch (error) {
        yield put(uploadFrameError("Save Frame failed"));
    }
}

export function* watchForUploadFrame() {
    yield takeLatest(FRAME_UPLOAD_REQUESTED, uploadFrame);
}

export function* getFrames(action) {
    try {
        const token = yield select(selectors.token);
        if (token) {
            const completePayload = Object.assign(action.payload, { token });
            const response = yield call(bike.getFrames, completePayload);
            yield put(getFrameListOK(response.data));
        } else {
            yield call(history.push, "/login");
        }
    } catch (error) {
        yield put(getFrameListError("Get Frames failed"));
    }
}

export function* watchForGetFrames() {
    yield takeLatest(FRAME_LIST_REQUESTED, getFrames);
}
