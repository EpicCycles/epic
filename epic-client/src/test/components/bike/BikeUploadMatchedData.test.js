import React from 'react';
import {BikeUploadMatchedData} from "../../../components/bike/BikeUploadMatchedData";

test("it renders correctly", () => {
    const matched = { rowIndex: 12, rowField: "Chain" };
    const component = shallow(<BikeUploadMatchedData
        matched={matched}
        matchIndex={0}
    />);
    expect(component).toMatchSnapshot();
});
test("it renders correctly when not first match", () => {
    const matched = { rowIndex: 12, rowField: "Chain" };
    const component = shallow(<BikeUploadMatchedData
        matched={matched}
        matchIndex={3}
    />);
    expect(component).toMatchSnapshot();
});