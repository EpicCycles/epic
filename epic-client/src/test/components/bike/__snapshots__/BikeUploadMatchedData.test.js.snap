// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly 1`] = `
<div
  className="rounded-auto"
  key="match120"
>
  Chain
  <Icon
    as="i"
    key="matchDelete$120"
    name="delete"
    onClick={[Function]}
  />
</div>
`;

exports[`it renders correctly when not first match 1`] = `
<div
  className="rounded-auto red"
  key="match123"
>
  Chain
  <Icon
    as="i"
    key="matchDelete$123"
    name="delete"
    onClick={[Function]}
  />
</div>
`;
